### PR TITLE
mycli: 1.6.0 -> 1.17.0

### PR DESCRIPTION
**Pull request by kalbasit**
###### Motivation for this change

I'm getting errors when using mycli 1.6.0, not getting errors after the update.

```python
Exception in thread Thread-4:
Traceback (most recent call last):
  File "/nix/store/7sbvn0wgv7hsnxpss1jba723kx5nz6d4-python-2.7.15/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/nix/store/7sbvn0wgv7hsnxpss1jba723kx5nz6d4-python-2.7.15/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/nix/store/8gs9jrpz8g9hmj4cz1326mr8k39y9nyw-python2.7-prompt_toolkit-1.0.15/lib/python2.7/site-packages/prompt_toolkit/interface.py", line 865, in run
    completions = list(buffer.completer.get_completions(document, complete_event))
  File "/nix/store/rsvy8bnpdninfzm9x3vk8l7a7vsgmzns-mycli-1.6.0/lib/python2.7/site-packages/mycli/sqlcompleter.py", line 255, in get_completions
    suggestions = suggest_type(document.text, document.text_before_cursor)
  File "/nix/store/rsvy8bnpdninfzm9x3vk8l7a7vsgmzns-mycli-1.6.0/lib/python2.7/site-packages/mycli/packages/completion_engine.py", line 85, in suggest_type
    full_text, identifier)
  File "/nix/store/rsvy8bnpdninfzm9x3vk8l7a7vsgmzns-mycli-1.6.0/lib/python2.7/site-packages/mycli/packages/completion_engine.py", line 134, in suggest_based_on_last_token
    token_v = token.value.lower()
AttributeError: 'tuple' object has no attribute 'value'
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

closes kalbasit/system#32

### CODE REVIEW

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -1102,6 +1102,20 @@ in {
 
   cheroot = callPackage ../development/python-modules/cheroot {};
 
+  cli-helpers = buildPythonPackage rec {
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: This needs to be put in `pkgs/development/python-modules/cli-helpers/default.nix`. The expression should adhere to the newest standards (`pname` etc.) and adding `cli-helpers` should be a separate commit.

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by dotlambda** [comment]
@kalbasit No need to have the second PR. Please remove the merge commit from this one.

@GrahamcOfBorg build python2.pkgs.cli-helpers python3.pkgs.cli-helpers

**Comment by dotlambda** [comment]
> Ran 0 tests in 0.001s

Please provide an appropriate `checkPhase` or set `doCheck = false` and add a comment stating the reason.

**Comment by kalbasit** [comment]
@dotlambda I edited the commits, the first one is #44578 verbatim and the second one is the primary commit of this PR.

**Comment by dotlambda** [comment]
But now you removed the commit regarding `pymysql` :confused: 

> you asked for this PR in an earlier comment

No, I just asked for a separate commit, no need to have multiple PRs.

**Comment by kalbasit** [comment]
oh, I see what you mean. Ok give me a second, I will split the last commit and close #44578.

**Comment by dotlambda** [comment]
Judging from `tox.ini`, it seems like they're using pytest, so the following might be enough
```
checkInputs = [ pytest ];
checkPhase = ''
  py.test
'';
```

**Comment by dotlambda** [comment]
@GrahamcOfBorg build python2.pkgs.cli-helpers python3.pkgs.cli-helpers

File: pkgs/tools/admin/mycli/default.nix [hunk 2]
```diff
@@ -1,28 +1,31 @@
 { lib
 , python
+, pythonPackages
```
Comment (dotlambda) [inline pkgs/tools/admin/mycli/default.nix]: Drop this line

File: pkgs/tools/admin/mycli/default.nix [hunk 3]
```diff
@@ -1,28 +1,31 @@
...
-    pymysql configobj sqlparse prompt_toolkit pygments click pycrypto
+    pymysql configobj sqlparse prompt_toolkit pygments click pycrypto cli-helpers
   ];
 
-  postPatch = ''
-    substituteInPlace setup.py --replace "==" ">="
-  '';
+  patches = [ ./fix-tests.patch ];
+
+  checkInputs = with pythonPackages; [ pytest mock ];
```
Comment (dotlambda) [inline pkgs/tools/admin/mycli/default.nix]: Drop `with pythonPackages;`. There's `with python.pkgs;` at the top.

File: pkgs/tools/admin/mycli/default.nix [hunk 4]
```diff
@@ -1,28 +1,31 @@
...
 
-  # No tests in archive. Newer versions do include tests
-  doCheck = false;
+  preCheck = ''
+    echo '[main]' > /tmp/myclirc
+    echo 'log_file = /tmp/mycli.test.log' >> /tmp/myclirc
```
Comment (dotlambda) [inline pkgs/tools/admin/mycli/default.nix]: From where would `myclirc` be read without the patch?

**Comment by kalbasit** [comment]
@dotlambda can you also build mycli to make sure it works on borg.

File: pkgs/tools/admin/mycli/default.nix [hunk 5]
```diff
@@ -1,28 +1,31 @@
 { lib
 , python
```
Comment (dotlambda) [inline pkgs/tools/admin/mycli/default.nix]: If mycli is compatible with python3, please replace this by `python3`. Otherwise, by `python2`.

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by dotlambda** [comment]
@GrahamcOfBorg build mycli

File: pkgs/tools/admin/mycli/default.nix (hunk 4 above)
Comment (kalbasit) [reply pkgs/tools/admin/mycli/default.nix]: From nowhere, and so the library will end up calling [write_default_config](https://github.com/dbcli/mycli/blob/dde8320b04f42345a1b83520ac646b2bc06ea229/mycli/main.py#L134-L135) which writes the config file at `/homeless-shelter/.myclirc` and that of course fails. Setting the myclirc stops it from writing the config file. Also, I specified the `log_file` option because another test invokes logging and that defaults to `/homeless-shelter/.mycli.log`.

Comment (dotlambda) [reply pkgs/tools/admin/mycli/default.nix]: How about setting something like `HOME=.`?

Comment (kalbasit) [reply pkgs/tools/admin/mycli/default.nix]: I'll try that!

Comment (kalbasit) [reply pkgs/tools/admin/mycli/default.nix]: This worked! Thanks for the tip.

File: pkgs/tools/admin/mycli/default.nix (hunk 5 above)
Comment (kalbasit) [reply pkgs/tools/admin/mycli/default.nix]: I tried using python3, the tests fail like this:

```sh-session
=================================== FAILURES ===================================
________________________________ test_list_dsn _________________________________

    def test_list_dsn():
        runner = CliRunner()
        with NamedTemporaryFile(mode="w") as myclirc:
            myclirc.write(dedent("""\
                [alias_dsn]
                test = mysql://test/test
                [main]
                log_file = /tmp/mycli.test.log
                """))
            myclirc.flush()
            args = ['--list-dsn', '--myclirc', myclirc.name]
            result = runner.invoke(cli, args=args)
>           assert result.output == "test\n"
E           AssertionError: assert '' == 'test\n'
E             + test

args       = ['--list-dsn', '--myclirc', '/build/tmpbldvl9dg']
myclirc    = <tempfile._TemporaryFileWrapper object at 0x7fffedb39da0>
result     = <Result RuntimeError('Click will abort further execution because Python 3 was configured to use ASCII as encoding for ... suitable UTF-8\nlocales were discovered.  This most likely requires resolving\nby reconfiguring the locale system.',)>
runner     = <click.testing.CliRunner object at 0x7fffedb39dd8>

test/test_main.py:287: AssertionError
==== 1 failed, 186 passed, 59 skipped, 1 xfailed, 1 xpassed in 5.42 seconds ====
builder for '/nix/store/m4vvpfnpgnlsa265cagv1hq99yfh5fs8-mycli-1.17.0.drv' failed with exit code 1
e
```

I put `python2` there for now, we can revisit in a separate PR.

**Comment by dotlambda** [comment]
@GrahamcOfBorg build python2.pkgs.pymysql python3.pkgs.pymysql mycli

File: pkgs/tools/admin/mycli/default.nix [hunk 6]
```diff
@@ -1,28 +1,29 @@
 { lib
-, python
+, python2
```
Comment (dotlambda) [inline pkgs/tools/admin/mycli/default.nix]: According to https://github.com/dbcli/mycli/blob/19619a4dfa0682654634c33e4ec55263f82475b6/tox.ini#L2, mycli supports python3.

File: pkgs/development/python-modules/pymysql/default.nix [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  version = "0.9.2";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0gvi63f1zq1bbd30x28kqyx351hal1yc323ckp0mihainb5n1iwy";
+  };
+
+  propagatedBuildInputs = [ cryptography ];
+
+  buildInputs = [ unittest2 ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/pymysql/default.nix]: I think this is not necessary.

File: pkgs/development/python-modules/pymysql/default.nix [hunk 8]
```diff
@@ -0,0 +1,31 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, python
```
Comment (dotlambda) [inline pkgs/development/python-modules/pymysql/default.nix]: Drop this line.

File: pkgs/tools/admin/mycli/default.nix (hunk 6 above)
Comment (kalbasit) [reply pkgs/tools/admin/mycli/default.nix]: I added a comment about it above, pasting here as well.

I tried using python3, the tests fail like this:

```sh-session
=================================== FAILURES ===================================
________________________________ test_list_dsn _________________________________

    def test_list_dsn():
        runner = CliRunner()
        with NamedTemporaryFile(mode="w") as myclirc:
            myclirc.write(dedent("""\
                [alias_dsn]
                test = mysql://test/test
                [main]
                log_file = /tmp/mycli.test.log
                """))
            myclirc.flush()
            args = ['--list-dsn', '--myclirc', myclirc.name]
            result = runner.invoke(cli, args=args)
>           assert result.output == "test\n"
E           AssertionError: assert '' == 'test\n'
E             + test

args       = ['--list-dsn', '--myclirc', '/build/tmpbldvl9dg']
myclirc    = <tempfile._TemporaryFileWrapper object at 0x7fffedb39da0>
result     = <Result RuntimeError('Click will abort further execution because Python 3 was configured to use ASCII as encoding for ... suitable UTF-8\nlocales were discovered.  This most likely requires resolving\nby reconfiguring the locale system.',)>
runner     = <click.testing.CliRunner object at 0x7fffedb39dd8>

test/test_main.py:287: AssertionError
==== 1 failed, 186 passed, 59 skipped, 1 xfailed, 1 xpassed in 5.42 seconds ====
builder for '/nix/store/m4vvpfnpgnlsa265cagv1hq99yfh5fs8-mycli-1.17.0.drv' failed with exit code 1
e
```

I put `python2` there for now, we can revisit in a separate PR.

File: pkgs/development/python-modules/pymysql/default.nix [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  buildInputs = [ unittest2 ];
+
+  # Wants to connect to MySQL
+  doCheck = false;
+
+  meta = with lib; {
+    description = "Pure Python MySQL Client";
+    homepage = https://github.com/PyMySQL/PyMySQL;
+    license = licenses.mit;
+    maintainers = [ maintainers.fridh maintainers.kalbasit ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/pymysql/default.nix]: Why did you add fridh?

Comment (kalbasit) [reply pkgs/development/python-modules/pymysql/default.nix]: I ran `git blame` and tracked down the first commit to him, should I remove him?

File: pkgs/tools/admin/mycli/default.nix (hunk 6 above)
Comment (dotlambda) [reply pkgs/tools/admin/mycli/default.nix]: You have to add `glibcLocales` to `checkInputs` and add `LC_ALL = "en_US.UTF-8";` to the expression.

File: pkgs/development/python-modules/pymysql/default.nix (hunk 9 above)
Comment (dotlambda) [reply pkgs/development/python-modules/pymysql/default.nix]: Yes, he's the main python maintainer but probably does not particularly maintain this package.

File: pkgs/tools/admin/mycli/default.nix (hunk 6 above)
Comment (kalbasit) [reply pkgs/tools/admin/mycli/default.nix]: Nice! it worked!

**Comment by kalbasit** [comment]
@dotlambda I simplified mycli even more, there really was no need for the patch, PTAL.

**Comment by dotlambda** [comment]
@GrahamcOfBorg build mycli
